botocore/client.py

```diff
@@ -325,6 +246,144 @@ def _api_call(self, *args, **kwargs):
         return _api_call
 
 
+class ClientEndpointBridge(object):
+    """Bridges endpoint data and client creation
+
+    This class handles taking out the relevant arguments from the endpoint
+    resolver and determining which values to use, taking into account any
+    client configuration options and scope configuration options.
+
+    This class also handles determining what, if any, region to use if no
+    explicit region setting is provided. For example, Amazon S3 client will
+    utilize "us-east-1" by default if no region can be resolved."""
+
+    SIGNATURE_PRECEDENCE = ['s3', 's3v4', 'v4', 'v2']
+    DEFAULT_ENDPOINT = '{service}.{region}.amazonaws.com'
+
+    def __init__(self, endpoint_resolver, scoped_config,
+                 client_config, default_endpoint=None):
+        self.endpoint_resolver = endpoint_resolver
+        self.scoped_config = scoped_config
+        self.client_config = client_config
+        self.default_endpoint = default_endpoint or self.DEFAULT_ENDPOINT
+
+    def resolve(self, service_name, region_name=None, endpoint_url=None,
+                is_secure=True):
+        region_name = self._check_default_region(service_name, region_name)
+        resolved = self.endpoint_resolver.construct_endpoint(
+            service_name, region_name)
+        if resolved:
+            return self._create_endpoint(
+                resolved, service_name, region_name, endpoint_url, is_secure)
+        else:
+            return self._assume_endpoint(service_name, region_name,
+                                         endpoint_url, is_secure)
+
+    def _check_default_region(self, service_name, region_name):
+        if region_name is not None:
+            return region_name
+        # Use the client_config region if no explicit region was provided.
+        if self.client_config and self.client_config.region_name is not None:
+            return self.client_config.region_name
+        # Use us-east-1 as the default region for S3.
+        if service_name == 's3':
+            return 'us-east-1'
```

Comment:
I'm hesitant to keep it in code, but I think if no one else feels that strongly about it, let's keep it.

As an aside, the only reason botocore's data files are called `_endpoints` and `_retry` are because back in the day, botocore didn't track service models by api version.  They were all in a single, top level directory by name, e.g: `['ec2.json', 's3.json', 'elb.json', ...]`.  Therefore, putting an `endpoints.json` in this same directory would make botocore think there's an "endpoints" service.  Even if the botocore loaders were hard coded to ignore `endpoints.json` from the `listdir()` call, we'd still be in trouble if an actual service called "endpoints" ever came out.  To workaround this, I just prefixed all non service model files with `_` and had the loaders ignore anything starting with `_`, giving us `_endpoints.json` and `_retry.json`.  This is no longer necessary now that each service is in its own directory, but we never went back and removed the leading `_` because there wasn't a real pressing reason to do so.  I never meant it as any kind of statement of internal/implementation detail.